File: ngram/nidhin_jose/ngram.c
/*
Write a program which first collects all word-level trigrams in a novel-length story. Then, choosing any arbitrary trigram 
as the starting point, generate a short story which is between 200 to 300 words. The next trigram in the chain must always 
have the first two words same as that of the previous trigrams last two words.

What is a trigram you ask? The sentence "The quick red fox jumps over the lazy brown dog" has the following trigrams - 
"The quick red", "quick red fox", "red fox jumps", "fox jumps over", "jumps over the", "over the lazy", "the lazy brown", 
"lazy brown dog".

Hint: Project Gutenberg is a good place to find a lot of plain text books
*/

/*NAME: NIDHIN JOSE*/

#include<stdio.h>

#define MAX_WORDLENGTH 100

void trigram_generator();
void shortstory();
int randomgenerator();

int main()
{
       trigram_generator();
       shortstory();
       printf("\n\nProgram successfully completed");
       
}

void shortstory()
{
     FILE *trigram_file,*story_file;
     int n,wordlimit;
     char c;
     char str[MAX_WORDLENGTH];
     trigram_file=fopen("trigram.txt","r");
    if(trigram_file==NULL)
    {
                        perror("Trigram_file");
                        exit(0);
                        }
    story_file=fopen("story.txt","w");
    if(story_file==NULL)
    {
                        perror("story_file");
                        exit(0);
                        }
    n=randomgenerator();
    wordlimit=n%100+200;
    fseek(trigram_file,0,SEEK_END);
    n=n% (ftell(trigram_file));
    if(n%2)
    fseek(trigram_file,n,SEEK_SET);
    else
    fseek(trigram_file,-n,SEEK_END);
    while(fscanf(trigram_file,"%c",&c))
    {
                                         if(c=='.')
                                         break;
                                               }
    fscanf(trigram_file,"%s",str);
    fscanf(trigram_file,"%s",str);
    for(n=0;n<wordlimit;n++)
    {
                            if(fscanf(trigram_file,"%s",str)!=1)
                            break;
                            fprintf(story_file,"%s ",str);
                            fscanf(trigram_file,"%s",str);
                            fscanf(trigram_file,"%s",str);
                            }
    fscanf(trigram_file,"%c",&c);
                           
    while(fscanf(trigram_file,"%c",&c))
    {
                                                   fprintf(story_file,"%c",c);
                                                   if(c=='.')
                                                   break;
                                                   else if (c==' ')
                                                   {
                                                        fscanf(trigram_file,"%s",str);
                                                        fscanf(trigram_file,"%s",str);
                                                        fscanf(trigram_file,"%c",&c);
                                                        fprintf(story_file,"%c",c);
                                                        }
                                                                                                     
                                                   }
           
    printf("\n\nShort story created in the file story.txt");                        
    fclose(trigram_file);
    fclose(story_file);
}
    
void trigram_generator()
{
     FILE *novel_file,*trigram_file;
    char str[3][MAX_WORDLENGTH];
    int n=3,a=4,b=5;
    novel_file=fopen("novel.txt","r");
    if(novel_file==NULL)
    {
                        perror("Novel_file");
                        exit(0);
                        }
    trigram_file=fopen("trigram.txt","w");
    if(trigram_file==NULL)
    {
                        perror("Trigram_file");
                        exit(0);
                        }
 
    if(fscanf(novel_file,"%s%s%s",str[0],str[1],str[2])!=3)
    {
                                                       printf("Not Enough content in the novel");
                                                       exit(0);
                                                       }
                                                       
    do
    {
                        fprintf(trigram_file,"%s %s %s ",str[n%3],str[a++%3],str[b++%3]); 
                        }while((fscanf(novel_file,"%s",str[n++%3])!=EOF));
    fclose(novel_file);
    fclose(trigram_file);
    printf("\n\nTrigram created in the trigram.txt");
    }
    
    int randomgenerator()
{
     srand ((unsigned int)time(NULL));
     return rand();    
}

File: ngram/renjith_p_ravindran/ngram.py
#!/usr/bin/python
#renjithforever@gmail.com | MOBME CODE JAM | NGRAM
"""
Contest ID: ngram
Write a program which first collects all word-level trigrams in a novel-length story. Then, choosing any arbitrary trigram as the starting point, generate a short story which is between 200 to 300 words. The next trigram in the chain must always have the first two words same as that of the previous trigrams last two words.
What is a trigram you ask? The sentence "The quick red fox jumps over the lazy brown dog" has the following trigrams - "The quick red", "quick red fox", "red fox jumps", "fox jumps over", "jumps over the", "over the lazy", "the lazy brown", "lazy brown dog".
"""

import string,random,re

if __name__=='__main__':
	input=open("input.txt",'r')
	input=input.read()
	input=re.sub("\W"," ",input)				# removes all punctuatiouns special chars etc.
	input=string.lower(input)
	
	input =input.split(" ")					# the string is now split into words and put into a python list
	
	inputT=[]
	for index in range(len(input)):				# performs some more cleansing on the input..could have done this with regex though :P
		if input[index]=='':
			continue
		elif len(input[index])==1 and input[index] !='a' and input[index]!='i':
			continue
		inputT.append(input[index])
	input=inputT

	trigram=[]
	i=0
	#trigramify!
	for index in range(len(input)-2):			#every trigram is found linerly
		trigram.append(input[index]+" "+input[index+1]+" "+input[index+2])

	"""
	two more trigrams are added so that the last trigrams seemlessly connect to the first trigram,
	when the trigrams are wrapped to front
	"""
	trigram.append(input[len(input)-2]+" "+input[len(input)-1]+" "+input[0])
	trigram.append(input[len(input)-1]+" "+input[0]+" "+input[1])

		
	position=random.randrange(0,len(trigram) - 1)			#a random psotion to start the story
#	position=len(trigram)-5
	#uncomment the above line to start the story from one of the last trigram and still get a 250 word story...it wraps to the first trigram..

	storyWordCount=0
	story=''
	while storyWordCount <=250:					#average of the range u ppl gave :)
		sentenceLength=random.randrange(4,12)			# choosing a random word lenght for every sentence
		sentence=trigram[position].capitalize()			#capitalizing the first word of the sentence
		sentenceWordCount=len(trigram[position].split(" "))
		i=0	 						#offset to `position` ie.. i th trigram form `position` th trigram 
		while sentenceWordCount <= sentenceLength: 		#adding more and more trigrams untlill sentenceLenght is reached
			i=i+1
			if position + i==len(trigram):			# when the last trigram is reached it is wraped to the first one.
				position=0
				i=0
			nextTrigram=trigram[position + i]
			sentenceWordCount+=len(trigram[position].split(" "))
			sentence+=" "+nextTrigram

		position=position+i+1

		storyWordCount+=len(sentence.split(" "))
		sentence+=". "						#finishing a sentence!
		
		story+=sentence						#adding the finished sentence to the story...
				
	print story
	#print len(story.split(" "))

